### DISCUSSION TITLE
How to fix a safetensors/ckpt model for a "state_dict" error

### DISCUSSION THREAD

**Discussion by jsbach-jung** [d8ahazard/sd_dreambooth_extension#794] [Show and tell]
When you try to train a safetensor, or a ckpt that has been merged with a safetensor, or one that was converted, you may see this:

```
RuntimeError: Error(s) in loading state_dict for UNet2DConditionModel:
Missing key(s) in state_dict: "up_blocks.0.upsamplers.0.conv.weight", "up_blocks.0.upsamplers.0.conv.bias", "up_blocks.1.upsamplers.0.conv.weight", "up_blocks.1.upsamplers.0.conv.bias", "up_blocks.2.upsamplers.0.conv.weight", "up_blocks.2.upsamplers.0.conv.bias".
Unexpected key(s) in state_dict: "up_blocks.0.attentions.2.conv.bias", "up_blocks.0.attentions.2.conv.weight".

```
Wanted to let you all know about my workaround, since I've seen a number of times now. Go to the Checkpoint Merger in the UI.

1. In the Primary slot, put in the SD1.5 model ckpt, the full 7-8GB one, or the NAI/Anything full one if you are training an anime model.
2. In the Secondary slot, put in your safetensors/broken ckpt model
3. Do a weighted sum of 1 by setting the slider all the way to the right and ensure the output is ckpt. I typically give it a suffix of -conv to indicate I did this to the model.
4. This converted model should be trainable, I have not had this process fail me even once.

The resulting ckpt file should now allow you to train on the model. I ran a few batches to ensure that the output result was visually similar. I'd say it's ~99.9% similar, you may notice slight differences. 

What's interesting is that putting the safetensors file in the primary slot and creating a ckpt with a WS value of 0 will still give you errors. Even doing a "full overwrite" with a WS value of 1 with the trainable SD1.5 ckpt will still produce errors.

There appears to be something about putting in a trainable ckpt model in as the Primary value and doing a full copy of the safetensor's values to it that keeps the conversion "clean".

Let me know if you have a theory on why this "fixes" the problem, I'm curious. And also if it solved the problem for  you.

**Comment by cschwarzhw** [comment]
Results in: Error merging checkpoints: unhashable type: 'list'

Error loading/saving model file:
Traceback (most recent call last):
  File "C:\stable-diffusion-webui\modules\ui.py", line 1690, in modelmerger
    results = modules.extras.run_modelmerger(*args)
  File "C:\stable-diffusion-webui\modules\extras.py", line 292, in run_modelmerger
    tertiary_model_info = sd_models.checkpoints_list.get(tertiary_model_name, None)
TypeError: unhashable type: 'list'

**Reply by Flonixcorn** [reply]
> Results in: Error merging checkpoints: unhashable type: 'list'
> 
> Error loading/saving model file: Traceback (most recent call last): File "C:\stable-diffusion-webui\modules\ui.py", line 1690, in modelmerger results = modules.extras.run_modelmerger(*args) File "C:\stable-diffusion-webui\modules\extras.py", line 292, in run_modelmerger tertiary_model_info = sd_models.checkpoints_list.get(tertiary_model_name, None) TypeError: unhashable type: 'list'

This is fixed in the new updates, this is an auto1111 error, update

**Comment by Flonixcorn** [comment]
This fixes it!

**Reply by AidedArt** [reply]
What fixed it?

**Comment by seemebreakthis** [comment]
Thanks !!  I am using https://colab.research.google.com/github/TheLastBen/fast-stable-diffusion/blob/main/fast-DreamBooth.ipynb and this method works perfectly to generate a model that can be dreambooth'ed.

**Comment by gilroff** [comment]
wow thanks!

**Comment by meunumerotim333** [comment]
Since the SD 1.5 7Gb ir a large file, and my B model is also a large file.. I'm getting CUDA error using Colab.

**Comment by 78Alpha** [comment]
I had to do something similar for one of mine, however, I actually blended together 3 models I trained to get rid of the issue layers. Merging 2 still has the error but the third with add difference and anything as model C worked out.

Could be a secondary method, the add difference.

**Comment by rodjjo** [comment]
I had the same error when I was using diffusers_stablediff_conversion. I fixed the conversion there:

https://github.com/ratwithacompiler/diffusers_stablediff_conversion/pull/6

I don't know if it's related to your problem

**Reply by underwaterbepis** [reply]
This seems correct, specifically making the changes here

https://github.com/ratwithacompiler/diffusers_stablediff_conversion/pull/6/commits/f3434631ad0e480e153c2c9ec02a28fafdda764a

Seems to fix the issue

**Comment by cleverestx** [comment]
Should I also bake in the standard VAE if I'm using the same good source as you in your example?

**Comment by Skettalee** [comment]
I am having a problem with a model i made, its a ckpt file and its saying its corrupt but i did not do anything to it and it worked one day and stopped the next.  
The errors that come up when i try to load it are : 
Calculating sha256 for G:\Sd\webui\models\Stable-diffusion\People\caseym_model.ckpt: f8779fb50efa10545267dbecde2f41ff4cbb18f8e0153d20eb60562e9a6a7146
Loading weights [f8779fb50e] from G:\Sd\webui\models\Stable-diffusion\People\caseym_model.ckpt
*** Error verifying pickled file from G:\Sd\webui\models\Stable-diffusion\People\caseym_model.ckpt
*** -----> !!!! The file is most likely corrupted !!!! <-----
*** You can skip this check with --disable-safe-unpickle commandline argument, but that is not going to help you.
***
    Traceback (most recent call last):
      File "G:\Sd\webui\modules\safe.py", line 83, in check_pt
        with zipfile.ZipFile(filename) as z:
      File "C:\Users\damie\AppData\Local\Programs\Python\Python310\lib\zipfile.py", line 1267, in __init__
        self._RealGetContents()
      File "C:\Users\damie\AppData\Local\Programs\Python\Python310\lib\zipfile.py", line 1334, in _RealGetContents
        raise BadZipFile("File is not a zip file")
    zipfile.BadZipFile: File is not a zip file

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "C:\Users\damie\AppData\Local\Programs\Python\Python310\lib\pickle.py", line 1245, in load_persid
        pid = self.readline()[:-1].decode("ascii")
    UnicodeDecodeError: 'ascii' codec can't decode byte 0x80 in position 63: ordinal not in range(128)

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "G:\Sd\webui\modules\safe.py", line 137, in load_with_extra
        check_pt(filename, extra_handler)
      File "G:\Sd\webui\modules\safe.py", line 104, in check_pt
        unpickler.load()
      File "C:\Users\damie\AppData\Local\Programs\Python\Python310\lib\pickle.py", line 1213, in load
        dispatch[key[0]](self)
      File "C:\Users\damie\AppData\Local\Programs\Python\Python310\lib\pickle.py", line 1247, in load_persid
        raise UnpicklingError(
    _pickle.UnpicklingError: persistent IDs in protocol 0 must be ASCII strings

---
changing setting sd_model_checkpoint to People\caseym_model.ckpt: AttributeError
Traceback (most recent call last):
  File "G:\Sd\webui\modules\options.py", line 140, in set
    option.onchange()
  File "G:\Sd\webui\modules\call_queue.py", line 13, in f
    res = func(*args, **kwargs)
  File "G:\Sd\webui\modules\initialize_util.py", line 170, in <lambda>
    shared.opts.onchange("sd_model_checkpoint", wrap_queued_call(lambda: sd_models.reload_model_weights()), call=False)
  File "G:\Sd\webui\modules\sd_models.py", line 741, in reload_model_weights
    state_dict = get_checkpoint_state_dict(checkpoint_info, timer)
  File "G:\Sd\webui\modules\sd_models.py", line 315, in get_checkpoint_state_dict
    res = read_state_dict(checkpoint_info.filename)
  File "G:\Sd\webui\modules\sd_models.py", line 301, in read_state_dict
    sd = get_state_dict_from_checkpoint(pl_sd)
  File "G:\Sd\webui\modules\sd_models.py", line 245, in get_state_dict_from_checkpoint
    pl_sd = pl_sd.pop("state_dict", pl_sd)
AttributeError: 'NoneType' object has no attribute 'pop'

ANYONE HAVE A CLUE how to read what may be wrong and if i can fix it?
